File: packages/enforcement_py/cognara_enforce/enforcement.py
"""Enforcement SDK: token+policy check with local verify, remote fallback, revocation listener."""

from __future__ import annotations

import uuid
from typing import Any, Callable

from . import api, jwt, policy, websocket_


def create_enforcement(
    api_url: str = "http://localhost:8000",
    api_key: str = "",
    fail_closed: bool = False,
    high_risk_actions: list[str] | None = None,
    on_log: Callable[[dict], None] | None = None,
    jwks_override: dict[str, Any] | None = None,
) -> object:
    """
    Create enforcement instance. Returns object with check(token, action_name, metadata) and close().
    """
    if high_risk_actions is None:
        high_risk_actions = ["send_email", "delete_user", "transfer_funds"]

    revoked_tokens: set[str] = set()
    cache: dict[str, dict] = {}

    def on_revoked(token_id: str) -> None:
        revoked_tokens.add(token_id)
        if on_log:
            on_log({"type": "revocation", "token_id": token_id})

    ws = websocket_.create_revocation_listener(api_url, on_revoked)

    class Enforcement:
        def check(
            self,
            token: str,
            action_name: str,
            metadata: dict[str, Any] | None = None,
        ) -> dict[str, Any]:
            """
            Check token and policy. Returns {allowed: bool, reason?: str, evidence_id: str}.
            """
            evidence_id = str(uuid.uuid4())

            def log(payload: dict) -> None:
                if on_log:
                    on_log({"evidence_id": evidence_id, **payload})

            try:
                # 1. Local validation
                local = jwt.verify_token_locally(token, api_url, jwks_override)
                if not local.get("valid"):
                    reason = local.get("reason", "invalid_token")
                    log({"action": action_name, "result": "blocked", "reason": reason})
                    return {"allowed": False, "reason": reason, "evidence_id": evidence_id}

                jti = local.get("jti", "")
                scopes = local.get("scopes", [])

                # 2. Revocation cache
                if jti in revoked_tokens:
                    log({"action": action_name, "result": "blocked", "reason": "token_revoked"})
                    return {"allowed": False, "reason": "token_revoked", "evidence_id": evidence_id}

                # 3. Remote fallback
                status = cache.get(jti, {}).get("status")
                if status is None:
                    remote = api.validate_token_remote(token, api_url, api_key)
                    if remote:
                        status = remote["status"]
                        cache[jti] = {
                            "status": status,
                            "subject": remote.get("subject", ""),
                            "scopes": remote.get("scopes", []),
                        }
                        if status == "revoked":
                            revoked_tokens.add(jti)
                            log({"action": action_name, "result": "blocked", "reason": "token_revoked"})
                            return {"allowed": False, "reason": "token_revoked", "evidence_id": evidence_id}
                    else:
                        if fail_closed and action_name in high_risk_actions:
                            log({
                                "action": action_name,
                                "result": "blocked",
                                "reason": "fail_closed_backend_unreachable",
                            })
                            return {
                                "allowed": False,
                                "reason": "fail_closed_backend_unreachable",
                                "evidence_id": evidence_id,
                            }

                # 4. Policy check
                policy_result = policy.check_policy(action_name, scopes)
                if not policy_result.get("allowed"):
                    reason = policy_result.get("reason", "policy_denied")
                    log({"action": action_name, "result": "blocked", "reason": reason})
                    return {"allowed": False, "reason": reason, "evidence_id": evidence_id}

                log({"action": action_name, "result": "allowed"})
                return {"allowed": True, "evidence_id": evidence_id}

            except Exception as err:
                if fail_closed and action_name in high_risk_actions:
                    log({"action": action_name, "result": "blocked", "reason": str(err)})
                    return {"allowed": False, "reason": str(err), "evidence_id": evidence_id}
                raise

        def close(self) -> None:
            ws.close()

    return Enforcement()

File: packages/enforcement/src/enforcement.js
'use strict';

const { randomUUID } = require('crypto');
const { createRevocationListener } = require('./websocket.js');
const { verifyTokenLocally } = require('./jwt.js');
const { validateTokenRemote } = require('./api.js');
const { checkPolicy } = require('./policy.js');

function resolveFailMode(actionName, failMode, failModePerAction, failClosed, highRiskActions) {
  if (failModePerAction != null && actionName in failModePerAction) {
    return failModePerAction[actionName];
  }
  if (failModePerAction != null) {
    return failMode;
  }
  if (failClosed && highRiskActions.includes(actionName)) {
    return 'fail_closed';
  }
  return 'fail_open';
}

function createEnforcement(options = {}) {
  const {
    apiUrl = 'http://localhost:8000',
    apiKey = '',
    failClosed = false,
    highRiskActions = ['send_email', 'delete_user', 'transfer_funds'],
    failMode = 'fail_open',
    failModePerAction = null,
    onLog = null,
    jwksOverride = null, // for testing: pass { keys: [...] } to bypass fetch
    _testRevokedTokens = null, // for testing: shared set to inject revocations
  } = options;

  const revokedTokens = _testRevokedTokens || new Set();
  const cache = new Map(); // jti -> { status, subject, scopes, cachedAt }

  const ws = createRevocationListener(apiUrl, (tokenId) => {
    revokedTokens.add(tokenId);
    if (onLog) onLog({ type: 'revocation', token_id: tokenId });
  });

  return {
    async check({ token, action }) {
      const evidenceId = randomUUID();
      const log = (payload) => {
        if (onLog) onLog({ evidence_id: evidenceId, ...payload });
      };

      try {
        // 1. Local validation: signature + TTL
        const local = await verifyTokenLocally(token, apiUrl, jwksOverride);
        if (!local.valid) {
          log({ action, result: 'blocked', reason: local.reason });
          return { allowed: false, reason: local.reason, evidence_id: evidenceId };
        }

        const { jti, scopes } = local;

        // 2. Revocation cache
        if (revokedTokens.has(jti)) {
          log({ action, result: 'blocked', reason: 'token_revoked' });
          return { allowed: false, reason: 'token_revoked', evidence_id: evidenceId };
        }

        // 3. Remote fallback if not cached
        let status = cache.get(jti)?.status;
        if (status === undefined) {
          const remote = await validateTokenRemote(token, apiUrl, apiKey);
          if (remote) {
            status = remote.status;
            cache.set(jti, {
              status: remote.status,
              subject: remote.subject,
              scopes: remote.scopes,
              cachedAt: Date.now(),
            });
            if (remote.status === 'revoked') {
              revokedTokens.add(jti);
              log({ action, result: 'blocked', reason: 'token_revoked' });
              return { allowed: false, reason: 'token_revoked', evidence_id: evidenceId };
            }
          } else {
            const mode = resolveFailMode(
              action?.name,
              failMode,
              failModePerAction,
              failClosed,
              highRiskActions,
            );
            if (mode === 'fail_closed') {
              log({ action, result: 'blocked', reason: 'fail_closed_backend_unreachable' });
              return {
                allowed: false,
                reason: 'fail_closed_backend_unreachable',
                evidence_id: evidenceId,
              };
            }
            if (mode === 'soft_fail') {
              log({ action, result: 'allowed', degraded_mode: 'soft_fail' });
            }
          }
        }

        // 4. Policy check
        const policy = checkPolicy(action, scopes);
        if (!policy.allowed) {
          log({ action, result: 'blocked', reason: policy.reason });
          return { allowed: false, reason: policy.reason, evidence_id: evidenceId };
        }

        // 5. Fail-closed: re-check revoked before returning allowed (in-flight revocation)
        if (revokedTokens.has(jti)) {
          log({ action, result: 'blocked', reason: 'token_revoked' });
          return { allowed: false, reason: 'token_revoked', evidence_id: evidenceId };
        }

        log({ action, result: 'allowed' });
        return { allowed: true, evidence_id: evidenceId };
      } catch (err) {
        const mode = resolveFailMode(
          action?.name,
          failMode,
          failModePerAction,
          failClosed,
          highRiskActions,
        );
        if (mode === 'fail_closed') {
          log({ action, result: 'blocked', reason: err.message });
          return {
            allowed: false,
            reason: err.message,
            evidence_id: evidenceId,
          };
        }
        throw err;
      }
    },

    close() {
      ws.close();
    },
  };
}

module.exports = { createEnforcement };
